### PR TITLE
Impac! permissions

### PR DESCRIPTION
**Pull request by cesar-tonnoir**
- [x] mention dependency to mnohub: https://github.com/maestrano/maestrano-hub/pull/818

- Modify `current_user` so ACL is attached to each organization
- Can pass the impersonator id to get the **impersonator's** access rights instead of the user's ones
- Add CanCan abilities to restrict the actions on Impac! controllers

### CODE REVIEW

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/dashboards_controller.rb [hunk 1]
```diff
@@ -28,15 +28,10 @@ def show
   # POST /mnoe/jpi/v1/impac/dashboards
   #   -> POST /api/mnoe/v1/users/1/dashboards
   def create
-    # TODO: dashboards.build breaks as dashboard.organization_ids returns nil, instead of an
-    #       empty array. (see MnoEnterprise::Impac::Dashboard #organizations)
-    # @dashboard = dashboards.build(dashboard_create_params)
-    # TODO: enable authorization
-    # authorize! :manage_dashboard, @dashboard
-    # if @dashboard.save
+    authorize! :create_impac_dashboards, dashboards.build(dashboard_create_params)
```
Comment (ouranos) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/dashboards_controller.rb]: we're doing a double instantiation here.
Could we have?
```ruby
@dashboard = dashboards.build(dashboard_create_params)
authorize! :create_impac_dashboards, @dashboard 
if @dashboard.save
  ...
end
```
Or is it not working as the old commented code is saying?

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/kpis_controller.rb [hunk 2]
```diff
@@ -49,16 +49,15 @@ def index
...
+  # TODO: nest alert in as a param, with the current user as a recipient.
   def create
     if params[:kpi][:widget_id].present?
       return render_not_found('widget') if widget.blank?
-      authorize! :manage_widget, widget
     else
       return render_not_found('dashboard') if dashboard.blank?
-      authorize! :manage_dashboard, dashboard
     end  
+    authorize! :create_impac_kpis, kpi_parent.kpis.build(kpi_create_params)
```
Comment (ouranos) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/kpis_controller.rb]: Same here, double instantiation

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb [hunk 3]
```diff
@@ -24,6 +24,8 @@ def index
   #  -> POST /api/mnoe/v1/dashboards/:id/widgets
   def create
     if widgets
+      authorize! :create_impac_widgets, widgets.build(widget_create_params)
```
Comment (ouranos) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb]: Double instantiation

File: core/app/models/mno_enterprise/impac/dashboard.rb [hunk 4]
```diff
@@ -22,13 +22,13 @@ def full_name
     end
 
     # Return all the organizations linked to this dashboard and to which the user has access
-    # If the dashboard is a template, return all the current user's organization
+    # If the dashboard is a template, return all the current user's organizations
     def organizations(org_list = nil)
       if org_list
         return org_list if dashboard_type == 'template'
-        org_list.to_a.select { |e| self.organization_ids.include?(e.uid) }
+        org_list.to_a.select { |e| organization_ids.include?(e.uid) || organization_ids.include?(e.id) }
```
Comment (ouranos) [inline core/app/models/mno_enterprise/impac/dashboard.rb]: Isn't in only `uids` (based on the `where` clause below)?

File: core/lib/mno_enterprise/concerns/models/ability.rb [hunk 5]
```diff
@@ -66,86 +66,69 @@ def initialize(user)
...
-  def impac_abilities(user)
-    can :manage_impac, MnoEnterprise::Impac::Dashboard do |dhb|
-      dhb.organizations.any? && dhb.organizations.all? do |org|
-        !!user.role(org) && ['Super Admin', 'Admin'].include?(user.role(org))
-      end
+  # Enables / disables Impac! Angular capabilities
+  def impac_abilities(orgs_with_acl)
+    can :create_impac_dashboards, MnoEnterprise::Impac::Dashboard do |d|
+      orgs = d.organizations(orgs_with_acl)
+      orgs.present? && orgs.all? { |org| org.acl.dig(:related, :dashboards, :create) }
```
Comment (ouranos) [inline core/lib/mno_enterprise/concerns/models/ability.rb]: `dig` not compatible with ruby < 2.3

File: core/lib/mno_enterprise/concerns/models/ability.rb [hunk 6]
```diff
@@ -66,86 +66,69 @@ def initialize(user)
     end
 
     #===================================================
-    # Impac
+    # Admin abilities
     #===================================================
-    impac_abilities(user)
+    admin_abilities(user)
```
Comment (ouranos) [inline core/lib/mno_enterprise/concerns/models/ability.rb]: Is the order change intentional?

admin_abilities were called last as there supposed to have precedence over everything:

> An ability rule will override a previous one
See: https://github.com/CanCanCommunity/cancancan/wiki/Ability-Precedence

I guess it doesn't matter in this case as the Impac! ACL is coming from MnoHub anyway?

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/dashboards_controller.rb (hunk 1 above)
Comment (cesar-tonnoir) [reply api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/dashboards_controller.rb]: Indeed it doesn't work as `build + save` does not seem to behave the same as `create`: some parameters (as the owner_id for example) are missing when the `save` Her-built request is sent to mno-hub, as a result, the record is not persisted correctly.

I've tried several things but always run into the same issues so I think we'd better let this code unchanged for now.

File: core/app/models/mno_enterprise/impac/dashboard.rb (hunk 4 above)
Comment (cesar-tonnoir) [reply core/app/models/mno_enterprise/impac/dashboard.rb]: no: ids are passed when the request is a `create` - this organizations uids/ids thing on the impac components are super confusing and should be refactored accross all the applications

File: core/lib/mno_enterprise/concerns/models/ability.rb (hunk 6 above)
Comment (cesar-tonnoir) [reply core/lib/mno_enterprise/concerns/models/ability.rb]: it wasn't intentional - reverting

File: core/app/models/mno_enterprise/impac/dashboard.rb [hunk 7]
```diff
@@ -44,10 +44,7 @@ def filtered_widgets_templates
     end
 
     def to_audit_event
-      {
-        name: name,
-        organization_id: (owner_type == 'Organization') ? owner_id : nil
-      }
+      { name: name }
```
Comment (ouranos) [inline core/app/models/mno_enterprise/impac/dashboard.rb]: @cesar-tonnoir didn't spot this one during code review. Why was this removed? I think that'd introduce a regression (see MNOE-861)

Comment (cesar-tonnoir) [reply core/app/models/mno_enterprise/impac/dashboard.rb]: @ouranos I think this is because after this change, owner_type should never be `Organization` again, so `organization_id` would always be nil - I figured out it would be more relevant to simply delete the field?...
